### ISSUE TITLE
Change kernel naming scheme

### ISSUE THREAD

**Issue by rdspring1** [NVIDIA/Fuser#778]
Currently, the kernel name is defined as `kernel + fusion_id_`. (See https://github.com/NVIDIA/Fuser/blob/main/csrc/executor.h#L224-L228)

This issue proposes that we use a more descriptive kernel name.
`kernel___[fusion_id_]___[pointwise, reduction, persistent, transpose, matmul scheduler type]___[gpu architecture]`

`fusion_id_` requires tracking a static count of fusions in `FusionExecutor` class, which causes complications with parallel compilation (See Issue #762) and serialization. 

Here are some options to replace `fusion_id_` in `FusionExecutor`.
1. Use the current number of fusions in `FusionCache` in python frontend.
2. Hash timestamp with universal hash function.

```cpp
// A simple universal hash function between [0, 21474836487] that shuffles numbers without collision.
auto hash_fn = [a = 714468, b = 205029](int64_t x) {
      auto linear_shift = a * x + b;
      auto lower_half = linear_shift & 0xFFFF;
      auto upper_half = linear_shift >> 16;
      auto rotate = (lower_half << 16) | upper_half;
      return rotate % 2147483648;
    };
```

**Comment by naftalimurgor** [comment]
Hello @rdspring1 
This seems a great one to take on, I'd like to take on this and submit a PR, please provide pointers on:
1.  Accessing `FusionCache` in the Python frontend
2. Setting things up in terms of development envrionment, building etc.

**Comment by naoyam** [comment]
@naftalimurgor Thanks for your interest. Currently these issues are just for our team as we are not yet ready to accept external contributions.

**Comment by kevinstephano** [comment]
I would like to replace `kernel` with `nvfuser` as the name prefix so our kernels are clearly identifiable in profiles.

**Comment by naftalimurgor** [comment]
> @naftalimurgor Thanks for your interest. Currently these issues are just for our team as we are not yet ready to accept external contributions.

Hey @naoyam 
It's okay. Totally understandable.

**Comment by jacobhinkle** [comment]
> Here are some options to replace `fusion_id_` in `FusionExecutor`.
> 
> 1. Use the current number of fusions in `FusionCache` in python frontend.
> 2. Hash timestamp with universal hash function.

If we are hashing, why not go ahead and use a 128-bit UUID (UUID7 or UUID4)? It is pretty much made for these cases where you have independent agents that need to generate IDs that are unique with high probability. In this case, they could be used as keys in a relational database of kernels which makes merging DBs easier.

Content-based hashing would be even better: we could hash the Fusion IR along with some enable/disable options that influence scheduling/codegen and the nvfuser version, so we'd have stable IDs that could be used to retrieve the kernel directly. The collision rate for a 256-bit hash should be negligible. Of course we'd need to be able to hash the `Fusion`, but maybe we could just serde it then hash the resulting buffer?

**Comment by naftalimurgor** [comment]
> @naftalimurgor Thanks for your interest. Currently these issues are just for our team as we are not yet ready to accept external contributions.

Hey @naoyam 

How do I join a the team? This is a really interesting project.

**Comment by naoyam** [comment]
> > @naftalimurgor Thanks for your interest. Currently these issues are just for our team as we are not yet ready to accept external contributions.
> 
> Hey @naoyam
> 
> How do I join a the team? This is a really interesting project.

I just referred to our NVIDIA internal team of nvFuser. We have an [open position](https://nvidia.wd5.myworkdayjobs.com/en-US/NVIDIAExternalCareerSite/job/Senior-Software-Engineer---Automated-Parallel-Programming_JR1966648) if you're interested.

**Comment by naftalimurgor** [comment]
Thank you! I'll give it a short

**Comment by zasdfgbnm** [comment]
Should we also add index type as part of the name?

**Comment by xwang233** [comment]
I'd vote for content-based hashing. If same kernels are **always** dumped with same file names consistently, it definitely helps tracking the development progress over time, and we may be able to maintain a database for that.

**Comment by naoyam** [comment]
Performance tracking should be at the fusion level rather than individual kernel levels, so having consistent kernel names does not always help. For example, a fusion may be currently segmented into multiple kernels, but in the future an improved scheduler may be able to avoid segmentation, thus generating just a single kernel. 

I think just having sequential names like what we currently do but consistent naming would be good enough.

**Comment by kevinstephano** [comment]
> I think just having sequential names like what we currently do but consistent naming would be good enough.

The issue neglects to mention we want to not use sequential numbered kernels because we could deserialize a cache of kernels and run them with a slightly different model and the sequence would no longer be valid but we can reuse already compiled kernels.

**Comment by naoyam** [comment]
Is there any specific problem with sequentially numbered kernels? In general, the kernel name shouldn't be used to indicate anything specific, so shouldn't any naming scheme just be fine as long as it's consistent?

If that's the case, I just prefer simple sequential naming. I'd prefer to play with just `kernel1.cu`, `kernel2.cu`, etc. rather than seemingly random kernel names that could change run-to-run due to different timestamps or something.

**Comment by kevinstephano** [comment]
While it is nice for debugging to have sequence numbers, if we keep the sequence numbers that means a set of kernels is only valid for a specific sequence of kernels. The sequence could change if we deserialize a model and use the deserialized kernels from one model with a new model. For instance, language models are often pre-trained and then the pretrained model is added into a new model that gets fine tuned where it is the same model plus some extra stuff bolted on for a specific task hence you could reuse the bulk of the original model but the exact sequence of the kernels might be slightly different. Therefore, it would be better if the kernel name was not specific to the sequence they are generated.

**Comment by naoyam** [comment]
I understand the original sequence would not have any meaning when used after deserialization, but is there any problem if the original kernels were generated with sequential naming?

**Comment by rdspring1** [comment]
The hash value should be consistent with context-based hashing. The kernel segments could be ordered sequentially.
e.g., `kernel_[original_fusion_hash]_[kernel_segment_id]_`

For internal development, would it be useful to add the git commit id to the kernel name?

**Comment by naoyam** [comment]
Having sequential IDs seems convenient, but I still don't know what the context-based hashing is. Is it something that is used to look up a saved kernel?

> For internal development, would it be useful to add the git commit id to the kernel name?

Personally, I don't think so. I often compare two kernels generated from the same fusion but with different branches, so having the same kernel name is actually more convenient.
